### PR TITLE
fix(vbank): be sure to persist nonce state in the KVStore

### LINKED ISSUE #8: wrong Block.Header.AppHash
@michaelfig writes:
> This "apphash" error indicates that we have some nondeterminism with the software upgrade.  We're working hard to look into it, but unfortunately all you really can do is stop your node, unsafe-reset-all and restart your node.

```
мая 07 15:14:15 goginn-pc systemd[1]: ag-chain-cosmos.service: Main process exited, code=exited, status=1/FAILURE
мая 07 15:14:15 goginn-pc systemd[1]: ag-chain-cosmos.service: Failed with result 'exit-code'.
мая 07 15:14:19 goginn-pc systemd[1]: ag-chain-cosmos.service: Scheduled restart job, restart counter is at 38.
мая 07 15:14:19 goginn-pc systemd[1]: Stopped Agoric Cosmos daemon.
мая 07 15:14:19 goginn-pc systemd[1]: Started Agoric Cosmos daemon.
мая 07 15:14:21 goginn-pc ag-chain-cosmos[36690]: Error: error during handshake: error on replay: wrong Block.Header.AppHash.  Expected 251A0E17277C24D3DD66BF83A57687A3B55A748A53C9892D40BAA0243DB29ACA, got 156DCBDF7B0D2A4FC4EC2C0D17B6410F9B1D4F5ED99E084E88257A7FD6B70108
```

### PR DESCRIPTION
**Pull request by michaelfig**
Fixes Agoric/testnet-notes#8, Refs Agoric/testnet-notes#11

The `x/vbank` module was not persisting its `lastNonce` state.  Thus, if a node restarted, the `lastNonce` would start from 0 again, and cause apphash problems when compared to blocks from other validators, once it resulted in a message send from the vat that received a `vbank` update.

This is likely the cause of many problems in the current testnet (appHash mismatches upon restart, resulting in low seed peer counts, which compounded bad network interconnectivity).

### CODE REVIEW

**Comment by michaelfig** [comment]
I've tested this by setting up a local chain, examining genesis.json (`"lastNonce": "0"`), running the chain and sending a bank tx to it, shutting down one chain node and exporting its state (`"lastNonce": "1"`), restarting the node, and verifying there is no appHash disagreement.

File: golang/cosmos/proto/agoric/vbank/vbank.proto [hunk 1]
```diff
@@ -22,10 +22,14 @@ message State {
...
+        [(gogoproto.nullable) = false, (gogoproto.castrepeated) = "github.com/cosmos/cosmos-sdk/types.Coins"
+        ];
 
     // rewardRate is the amount of reward, if available, to send to the
     // fee collector module on every block.
     repeated cosmos.base.v1beta1.Coin rewardRate = 2
         [(gogoproto.nullable) = false, (gogoproto.castrepeated) = "github.com/cosmos/cosmos-sdk/types.Coins"];
+        
+    // lastNonce is a sequence number for communicating with the VM.
+    uint64 lastNonce = 3;
```
Comment (JimLarson) [inline golang/cosmos/proto/agoric/vbank/vbank.proto:34]: Call it `lastSeqence` then, here and below.

File: golang/cosmos/x/vbank/vbank_test.go [hunk 2]
```diff
@@ -476,10 +493,17 @@ func Test_EndBlock_Events(t *testing.T) {
 	if len(msgsSent) != 1 {
 		t.Errorf("got msgs = %v, want one message", msgsSent)
 	}
-	gotMsg, err := decodeBalances([]byte(msgsSent[0]))
+	gotMsg, gotNonce, err := decodeBalances([]byte(msgsSent[0]))
 	if err != nil {
 		t.Errorf("decode balances error = %v", err)
```
Comment (JimLarson) [inline golang/cosmos/x/vbank/vbank_test.go:498]: Make this `Fatalf` then - if we get here then `gotMsg` and `gotNonce` are likely garbage and we shouldn't compare further.

File: golang/cosmos/x/vbank/keeper/keeper.go [hunk 3]
```diff
@@ -39,23 +38,6 @@ func NewKeeper(
 	}
 }
 
-func (k Keeper) GetGenesis(ctx sdk.Context) types.GenesisState {
```
Comment (JimLarson) [inline golang/cosmos/x/vbank/keeper/keeper.go:42]: Double-check that you recall why this was added to confirm that it's safe to remove, and for `SetGenesis()` too.

**Review by JimLarson — APPROVED** [review]

File: golang/cosmos/x/vbank/keeper/keeper.go (hunk 3 above)
Comment (michaelfig) [reply golang/cosmos/x/vbank/keeper/keeper.go:42]: Yeah, it predated `Params` and its only purpose was to persist them for later export.

File: golang/cosmos/proto/agoric/vbank/vbank.proto (hunk 1 above)
Comment (michaelfig) [reply golang/cosmos/proto/agoric/vbank/vbank.proto:34]: Created and assigned #3424 to you.

File: golang/cosmos/x/vbank/vbank_test.go (hunk 2 above)
Comment (michaelfig) [reply golang/cosmos/x/vbank/vbank_test.go:498]: Created and assigned #3425 to you.

**Pull request merged**